impl/src/main/java/org/ehcache/internal/events/CacheEventNotificationListenerServiceProviderImpl.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright Terracotta, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ehcache.internal.events;
+
+import org.ehcache.events.CacheEventNotificationListenerServiceProvider;
+import org.ehcache.events.CacheEventNotificationService;
+import org.ehcache.events.CacheEventNotificationServiceImpl;
+import org.ehcache.events.DisabledCacheEventNotificationService;
+import org.ehcache.spi.ServiceProvider;
+import org.ehcache.spi.service.ServiceConfiguration;
+import org.ehcache.spi.service.ThreadPoolsService;
+
+/**
+ * @author palmanojkumar
+ *
+ */
+public class CacheEventNotificationListenerServiceProviderImpl implements CacheEventNotificationListenerServiceProvider {
+
+  private ServiceConfiguration<?> config;
+  private ServiceProvider serviceProvider;
```

Comment:
Right, but what provides the visibility guarantees to `. createCacheEventNotificationService` that access the `serviceProvider` field? 

hint:I actually think there _is_ one, but I have not verified whether it's entirely correct and further more, if you do see which it is, you'd probably agree it's _very_ brittle.